crates/hir-ty/benches/incremental.rs

```diff
@@ -0,0 +1,212 @@
+    }
+
+    fn set_source_root_with_durability(
+        &mut self,
+        source_root_id: SourceRootId,
+        source_root: Arc<SourceRoot>,
+        durability: Durability,
+    ) {
+        let files = Arc::clone(&self.files);
+        files.set_source_root_with_durability(self, source_root_id, source_root, durability);
+    }
+
+    fn file_source_root(&self, id: base_db::FileId) -> FileSourceRootInput {
+        self.files.file_source_root(id)
+    }
+
+    fn set_file_source_root_with_durability(
+        &mut self,
+        id: base_db::FileId,
+        source_root_id: SourceRootId,
+        durability: Durability,
+    ) {
+        let files = Arc::clone(&self.files);
+        files.set_file_source_root_with_durability(self, id, source_root_id, durability);
+    }
+
+    fn crates_map(&self) -> Arc<CratesMap> {
+        self.crates_map.clone()
+    }
+}
+
+#[salsa_macros::db]
+impl salsa::Database for BenchmarkDB {}
+
+impl panic::RefUnwindSafe for BenchmarkDB {}
+
+impl BenchmarkDB {
+    pub(crate) fn module_for_file_opt(&self, file_id: impl Into<FileId>) -> Option<ModuleId> {
+        let file_id = file_id.into();
+        for &krate in self.relevant_crates(file_id).iter() {
+            let crate_def_map = crate_def_map(self, krate);
+            for (local_id, data) in crate_def_map.modules() {
+                if data.origin.file_id().map(|file_id| file_id.file_id(self)) == Some(file_id) {
+                    return Some(crate_def_map.module_id(local_id));
+                }
+            }
+        }
+        None
+    }
+
+    pub(crate) fn module_for_file(&self, file_id: impl Into<FileId>) -> ModuleId {
+        self.module_for_file_opt(file_id.into()).unwrap()
+    }
+}
+
+fn benchmark_incremental_use_expr_addition(c: &mut Criterion) {
+    c.bench_function("incremental_use_expr_addition", |b| {
+        b.iter_batched(
+            || {
+                let (db, pos) = BenchmarkDB::with_position(
+                    "
+//- /lib.rs
+fn foo() -> i32 {
+    $01 + 1
+}",
+                );
+
+                let module = db.module_for_file(pos.file_id.file_id(&db));
+                let _crate_def_map = module.def_map(&db);
+                db.trait_impls_in_crate(module.krate());
+
+                (db, pos)
+            },
+            |(mut db, pos)| {
+                let new_text = "
+use std::collections::HashMap;
+
+fn foo() -> i32 {
+    $01 + 1
+}";
+
+                db.set_file_text(pos.file_id.file_id(&db), new_text);
+                let module = db.module_for_file(pos.file_id.file_id(&db));
+                let _crate_def_map = module.def_map(&db);
+
+                black_box(db.trait_impls_in_crate(module.krate()));
+            },
+            criterion::BatchSize::LargeInput,
+        );
+    });
+}
+
+fn benchmark_incremental_struct_addition(c: &mut Criterion) {
+    c.bench_function("incremental_struct_addition", |b| {
+        b.iter_batched(
+            || {
```

Comment:
I'm afraid the cost of the setup will dwarf the cost of the benchmarked computation, not allowing criterion to accurately measure the time.

Reply:
Oh, I’m sorry: I misunderstood. I have noticed this is somewhat noisy, but I chalked it up to my work-managed computer being bad.